docs/dev/index.rst

```diff
@@ -15,28 +15,360 @@
+
+`runtime::Object` is one of the primary data structures in TVM runtime besides the `runtime::PackedFunc`.
+It is a reference-counted base class with type index to support runtime type checking and downcasting.
+The object system allows the developer to introduce new data structures to the runtime, such as Array, Map, and new IR data structures.
+
+Besides the deployment use-cases, the compiler itself also makes heavy use of the TVM's runtime mechanism.
+All of the IR data structures are subclasses of `runtime::Object`, as a result, they can be directly accessed and manipulated from the Python frontend.
+We use the PackedFunc mechanism to expose various APIs to the frontend.
+
+Runtime support for different hardware backends are defined in subdirectories of runtime(e.g. runtime/opencl).
+These hardware-specific runtime modules define APIs for device memory allocation and device function serialization.
+
+`runtime/rpc` implements an RPC support for PackedFunc. We can use the RPC mechanism to send a cross-compiled library to a remote
+device and benchmark the execution performance. The rpc infrastructure enables data collection from a wide range of hardware backends
+for learning-based optimizations.
 
-Building a compiler stack for deep learning systems involves many many systems-level design decisions.
-In this part of documentation, we share the rationale for the specific choices made when designing TVM.
 
 .. toctree::
-   :maxdepth: 2
+   :maxdepth: 1
 
    runtime
    debugger
+   virtual_machine
+   introduction_to_module_serialization
+
+tvm/node
+--------
+The node module adds additional features on top of the `runtime::Object` for IR data structures.
+The main features include reflection, serialization, structural equivalence, and hashing.
+
+Thanks to the node module, we can directly access any field of the tvm's IRNode by their name in python.
+
+.. code-block:: python
+
+    x = tvm.tir.Var("x", "int32")
+    y = tvm.tir.Add(x, x)
+    # we can directly use the field name to access the IR structures
+    assert y.a == x
+
+
+We can also directly serialize arbitrary IR node into a JSON format, and load them back.
+The ability to save/store, and inspect an IR node provides a foundation for making the compiler more accessible.
+
+
+tvm/ir
+------
+The `tvm/ir` folder contains the unified data structure and interfaces across for all IR function variants.
+The components in `tvm/ir` are shared by `tvm/relay` and `tvm/tir`, notable ones include
+
+- IRModule
+- Type
+- PassContext and Pass
+- Op
+
+Different variants of functions(e.g. relay.Function and tir.PrimFunc) can co-exist in an IRModule.
+While these variants may not have the same content representation, they use the same data structure to represent types,
+and as a consequence, their function signatures. The unified type system allows one function variant to call into another
```

Comment:
- Didn't understand this: `and as a consequence, their function signatures`.
- s/call into/call/